From 6b645c10268341df439c0e85303b2780b6fe9946 Mon Sep 17 00:00:00 2001
From: Jens Rehsack <sno@NetBSD.org>
Date: Sun, 8 Jan 2012 13:19:42 +0100
Subject: [PATCH 4/4] rename cache variable for dirent header

---
 lib/Config/AutoConf.pm |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/lib/Config/AutoConf.pm b/lib/Config/AutoConf.pm
index 4e75c94..6f98500 100644
--- a/lib/Config/AutoConf.pm
+++ b/lib/Config/AutoConf.pm
@@ -1236,7 +1236,7 @@ might not need touse this macro.
 sub check_dirent_header {
   my $self = shift->_get_instance();
 
-  my $cache_name = $self->_cache_name( "dirent_with_DIRP" );
+  my $cache_name = $self->_cache_name( "header_dirent" );
   my $check_sub = sub {
   
     my $have_dirent;
-- 
1.7.7.2

